### PR TITLE
CORE-1462: url-encode paths for anonymous access

### PR DESCRIPTION
**Pull request by ianmcorvidae**
A little bit annoying, but this seems to work with what I've tried it with.

The main annoyance is that java.net.URLEncoder turns spaces into `+` characters. However, since it turns `+` into `%2B`, we can just do a string replace afterwards to get what we want, hacky though it feels. Other than the space-plus thing, it does what we want. Of course, we also don't want to encode the slashes, so I split on those and then re-join the path on them.

It probably won't matter either way, but I've set it up to do this after mapping the path to the appropriate webdav URL. It won't come up with anything we're running, but it could ostensibly matter if we wanted to map something with a space or other needing-url-encoding character to a different path before returning it; this way, the config file uses the unencoded versions, not the encoded ones.

### CODE REVIEW

**Review by sri2k1us — APPROVED** [review]
LGTM. (am also a novice when it comes to clojure so)

**Review by slr71 — APPROVED** [review]
Looks good to me! :shipit:

File: src/data_info/services/sharing.clj [hunk 1]
```diff
@@ -160,10 +161,15 @@
...
+(defn- encode-mapped-anon-path
+  "Take a path that's been appropriately mapped, and URL encode it for use. This function will use %20 for spaces."
+  [mapped-path]
+  (string/join "/" (map #(string/replace (URLEncoder/encode %) #"\+" "%20") (string/split mapped-path #"/"))))
+
 (defn anon-file-url
   [path]
   (let [aurl (url/url (cfg/anon-files-base-url))]
-    (str (-> aurl (assoc :path (apply ft/path-join (:path aurl) [(map-anon-url-path path)]))))))
+    (str (-> aurl (assoc :path (apply ft/path-join (:path aurl) [(encode-mapped-anon-path (map-anon-url-path path))]))))))
```
Comment (psarando) [inline src/data_info/services/sharing.clj:172]: Maybe it's just me, but the `->` threading only used for the `assoc` is making this hard to follow for me. I would go all out and thread everything, or not use threading at all here 😕 
```suggestion
    (-> aurl :path (ft/path-join (encode-mapped-anon-path (map-anon-url-path path))) (assoc :path))))
```

File: src/data_info/services/sharing.clj [hunk 2]
```diff
@@ -160,10 +161,15 @@
     (when-not (nil? matching-key)
       (string/replace-first path matching-key (mappings matching-key)))))
 
+(defn- encode-mapped-anon-path
+  "Take a path that's been appropriately mapped, and URL encode it for use. This function will use %20 for spaces."
+  [mapped-path]
+  (string/join "/" (map #(string/replace (URLEncoder/encode %) #"\+" "%20") (string/split mapped-path #"/"))))
```
Comment (psarando) [inline src/data_info/services/sharing.clj:167]: Looks good to me. I might have used `as->` threading here, but this is also easy enough to follow, I think.

**Review by psarando — APPROVED** [review]
LGTM 👍

File: src/data_info/services/sharing.clj (hunk 1 above)
Comment (psarando) [reply src/data_info/services/sharing.clj:172]: I'm not sure if my suggestion is actually correct 🤔

Comment (ianmcorvidae) [reply src/data_info/services/sharing.clj:172]: Hm, good point. I tried to change it as little as possible, as I generally do, but it's true it might be nice to thread it or otherwise break it up a bit more. It's certainly not written the most clearly it could be.

I guess the completely unthreaded version would be:

```clojure
(str
  (assoc aurl
         (apply ft/path-join
                (:path aurl)
                [(encode-mapped-anon-path
                   (map-anon-url-path path))])))
```
So I guess we could only thread `aurl` once more, into the `str` call, but maybe we could do better otherwise. I think using `apply` is extraneous, at least?

Maybe:

```clojure
(defn anon-file-url
  [path]
  (let [aurl (url/url (cfg/anon-files-base-url))
        encoded-path (encoded-mapped-anon-path (map-anon-url-path path))]
     (-> aurl
         (assoc :path (ft/path-join (:path aurl) encoded-path))
         str)))
```
... if I haven't typoed something there, anyway?

Comment (psarando) [reply src/data_info/services/sharing.clj:172]: Nice, LGTM! Even just moving the `encoded-path` into the `let` helps out 👍

File: src/data_info/services/sharing.clj (hunk 2 above)
Comment (ianmcorvidae) [reply src/data_info/services/sharing.clj]: I like that idea, I switched it over to using that.

**Comment by ianmcorvidae** [comment]
Thanks for the reviews, everyone!
